sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala

```diff
@@ -159,6 +159,10 @@ case class HashAggregateExec(
   // don't need a stop check before aggregating.
   override def needStopCheck: Boolean = false
 
+  // Aggregate operator always consumes all the input rows before outputting any result, so its
+  // upstream operators can keep producing data, even if there is a limit after Aggregate.
```

Comment:
I have not looked at this clearly. But if there is limit before Aggregate? We should not consume all input rows.

Reply:
let's say the query is `range -> limit -> agg -> limit`.

So `agg` does consume all the inputs, from the first `limit`. The range will have a stop check w.r.t. to first `limit`, not the second `limit`. If there is no limit before `agg`, then `range` will not have a stop check.